### ISSUE TITLE
More expressive ways to express compound variants

### ISSUE THREAD

**Issue by ethanresnick** [modulz/stitches#802]
**Is your feature request related to a problem? Please describe.**
Here's an example:

I have a `Button` component that has different `size` variants, which control the text size, padding, etc. Each `size` variant sets a different `borderRadius`, where bigger sizes get a bigger radius, but the size of the radius isn't exactly proportional to the text size (so it can't just be one constant value [e.g., in ems], and does have to be different for each variant). 

Now, I also want to have a boolean variant — `unrounded` — that sets `borderRadius: 0`, regardless of the size. My sense is that the only way to do this today is to make a compound variant for each (size, unrounded) pair, as in:

```ts
{
  compoundVariants: [
    {
      size: "small",
      unrounded: true
      css: {
        borderRadius: 0,
      },
    },
        {
      size: "medium",
      unrounded: true
      css: {
        borderRadius: 0,
      },
    },    {
      size: "large",
      unrounded: true
      css: {
        borderRadius: 0,
      },
    },    {
      size: "xl",
      unrounded: true
      css: {
        borderRadius: 0,
      },
    },
  ];
}
```

That feels quite clunky/verbose.

**Describe the solution you'd like**
In thinking about this, I'm realizing that I don't fully understand the current stitches API contract when multiple variants with conflicting styles apply, and there is no `compoundVariant` defining their interaction. That is, if I'd just defined `unrounded` as a normal, boolean variant that set `borderRadius` to 0 [with no `compoundVariant`], then the `Button` would have two conflicting `borderRadius` values — one from the `size` and one from `unrounded` — and I'm not sure if there's a current mechanism in the stitches API for controlling which declaration should override the other. If there is, then that would solve this use case. But, since I didn't see it documented, my guess is that the behavior in such a case is intentionally undefined, and that strikes me as the right approach, since overrides and precedence rules are much harder to learn and work with reliably than the explicit, opt-in matching of `compoundVariants`.

So I'd like some approach that makes it easier to express multiple combinations of variants in a more concise way. Here are some possibilities:

```ts
// accept an array of values for variants in the compound variant. Seems like a simple, logical extension of the current API.
{ unrounded: true, size: ["small", "medium", "large", "xl"], css: { borderRadius: 0 } }

// define some sort of "wild card" token so that all the values for size don't have to be enumerated explicitly/kept in sync.
// `any` would be a symbol exposed by the stitches package.
{ unrounded: true, size: any, css: { borderRadius: 0 } }

// the wildcard could also be a string, which saves an import and is maybe more intuitive 
// (leveraging common meaning of '*'), though this is going down the road of turning the string value
// into some variant-value matching grammar, which could be fine
{ unrounded: true, size: "*", css: { borderRadius: 0 } }

// Finally -- and idk the implementation implications of this, though i can imagine that it might make 
// certain optimizations harder -- users could provide a predicate fn to indicate if the compound variant matches.
{ test: (variants) => variants.unrounded, css: { borderRadius: 0 } }
```

**Describe alternatives you've considered**
See above.

**Additional context**
No

**Comment by peduarte** [comment]
Im not sure I fully understand the concern. Im probably missing some context. A reproduction would be useful.

My I'm guessing the source of the problem is the fact that you are adding border radius in the size variant. So then you're forced to remove the border radius via an `unrounded` variant. 

Ideally, adding CSS properties is easier than removing them. 

You could consider a slightly different approach:
- `size` variant only deals with size, no border radius
- by default, there are no border radius
- introduce a `shape/corner` variant, where it can be `pill/round` or `square/sharp` - up to you
- set the default variant that you use most often (optional)
- use `compoundVariants` to change border radius size based on `size` variant

```js
const Button = styled('button', {
  variants: {
    size: {
      1: { fontSize:  14 },
    },
    shape: {
      pill: {} 
    }
  },
  compoundVariants: [{
    size: '1',
    shape: 'pill',
    css: { borderRadius: 3 }
  }]
})
```

Then:
```jsx
// rounded
<Button size="1" shape="pill" />

// not rounded
<Button size="1" shape="square" />
```

You could also use a boolean variant for pill, but I try to use enum as much as possible

**Comment by ethanresnick** [comment]
> My I'm guessing the source of the problem is the fact that you are adding border radius in the size variant. So then you're forced to remove the border radius via an unrounded variant.

You're right: applying a radius as part of the size variant's styles was complicating things a lot. I was porting this component from an old design system and not thinking enough about how to refactor it.

Even making radius totally orthogonal to size, though, I am still ending up with a bit of repetition in my `compoundVariants`. To borrow your naming scheme, I actually have a `shape` variant with three values: `"square" | "rounded-rect" | "pill"`; at `rounded-rect` shape, the radius is always set, but some sizes use the same value, so I have something like:

```js
[{
  shape: "rounded-rect",
  size: "small",
  css: {
    borderRadius: "2px",
  },
},
{
  shape: "rounded-rect",
  size: "normal",
  css: {
    borderRadius: "2px",
  },
},
{
  shape: "rounded-rect",
  size: "large",
  css: {
    borderRadius: "3px",
  },
},
{
  shape: "rounded-rect",
  size: "xl",
  css: {
    borderRadius: "3px",
  },
}]
```

The repetition there is really just a result of me clipping the border radius values to values from our design system. I.e., conceptually, maybe the rounding on the normal size button should be `2.5px` or something, but we instead make it 2px (matching the `small` variant's rounding) since we don't otherwise use 2.5px border radii.

It'd be nice to be able to instead write:

```js
[{
  shape: "rounded-rect",
  size: ["small", "normal"],
  css: {
    borderRadius: "2px",
  },
},
{
  shape: "rounded-rect",
  size: ["large", "xl"]
  css: {
    borderRadius: "3px",
  },
}]
```

And an extension like that of the compoundVariants syntax really does seem to me like it would be generally useful.

**Comment by peduarte** [comment]
Yeah, that is by design though. Repetition FTW! This is one of the core values of Stitches. It is under-abstracted on purpose.
